test/integration/concierge_tls_spec_test.go

```diff
@@ -0,0 +1,243 @@
+		},
+		{
+			name: "should disallow certificate authority data source with invalid kind",
+			customResourceYaml: here.Doc(`
+			---
+			apiVersion: authentication.concierge.%s/v1alpha1
+			kind: WebhookAuthenticator
+			metadata:
+				name: %s
+			spec:
+				endpoint: "https://web-auth/token"
+				tls:
+					certificateAuthorityDataSource:
+						kind: sorcery
+						name: foo
+						key: bar
+			`),
+			customResourceName: "invalid-webhook-auth-invalid-kind",
+			expectedError:      `The WebhookAuthenticator "%s" is invalid: spec.tls.certificateAuthorityDataSource.kind: Unsupported value: "sorcery": supported values: "Secret", "ConfigMap"`,
+		},
+		{
+			name: "should create a custom resource passing all validations using a Secret source",
+			customResourceYaml: here.Doc(`
+			---
+			apiVersion: authentication.concierge.%s/v1alpha1
+			kind: WebhookAuthenticator
+			metadata:
+				name: %s
+			spec:
+				endpoint: "https://web-auth/token"
+				tls:
+					certificateAuthorityDataSource:
+						kind: Secret
+						name: foo
+						key: bar
+		`),
+			customResourceName: "valid-webhook-auth-secret-kind",
+			expectedError:      "",
+		},
+		{
+			name: "should create a custom resource passing all validations using a ConfigMap source",
+			customResourceYaml: here.Doc(`
+			---
+			apiVersion: authentication.concierge.%s/v1alpha1
+			kind: WebhookAuthenticator
+			metadata:
+				name: %s
+			spec:
+				endpoint: "https://web-auth/token"
+				tls:
+					certificateAuthorityDataSource:
+						kind: ConfigMap
+						name: foo
+						key: bar
+			`),
+			customResourceName: "valid-webhook-auth-cm-kind",
+			expectedError:      "",
+		},
+		{
+			name: "should create a custom resource without any tls spec",
+			customResourceYaml: here.Doc(`
+			---
+			apiVersion: authentication.concierge.%s/v1alpha1
+			kind: WebhookAuthenticator
+			metadata:
+				name: %s
+			spec:
+				endpoint: "https://web-auth/token"
+			`),
+			customResourceName: "no-tls-spec",
+			expectedError:      "",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			yamlFilepath := filepath.Join(t.TempDir(), fmt.Sprintf("tls-spec-validation-%s.yaml", tc.customResourceName))
+
+			resourceName := tc.customResourceName + "-" + testlib.RandHex(t, 7)
+			yamlBytes := []byte(fmt.Sprintf(tc.customResourceYaml, env.APIGroupSuffix, resourceName))
+
+			require.NoError(t, os.WriteFile(yamlFilepath, yamlBytes, 0600))
+			cmd := exec.CommandContext(context.Background(), "kubectl", []string{"apply", "-f", yamlFilepath}...)
+			var stdOut, stdErr bytes.Buffer
+			cmd.Stdout = &stdOut
+			cmd.Stderr = &stdErr
+
+			err := cmd.Run()
+			if tc.expectedError == "" {
+				require.NoError(t, err)
+				require.Equal(t, fmt.Sprintf("webhookauthenticator.authentication.concierge.pinniped.dev/%s created\n", resourceName), stdOut.String())
+				require.Empty(t, stdErr.String())
+				t.Cleanup(func() {
```

Comment:
Seems like it would be safer to use `t.Cleanup` immediately after calling `kubectl apply` to be sure that we will attempt to clean up no matter what happens with the other assertions and no matter which way it branches at`if tc.expectedError == ""`. You could use `kubectl delete --ignore-not-found=true` in the `t.Cleanup`, and then it would not matter if the resource was successfully created or not by `kubectl apply`.

Reply:
The test is attempting to delete the resource only when the resource creation w/ `kubectl apply` is successful. This happens only `tc.expectedError == ""` evaluates to true. Attempting to cleanup outside the `true` branch of the condition will have the following implication: